src/api/c/sparse.cpp

```diff
@@ -0,0 +1,395 @@
+
+    } else if(storage == AF_SPARSE_CSC) {
+        dense2storage<T, AF_SPARSE_CSC>(values, rowIdx, colIdx, in);
+    }
+
+    out->rowIdx = getHandle(rowIdx);
+    out->colIdx = getHandle(colIdx);
+    out->values = getHandle(values);
+
+}
+
+template<typename T>
+void create_sparse_array_from_dense1(af_sparse_t *out, const af_array in,
+                                     const af_array nonZeroIdx, const af_sparse_storage storage)
+{
+    switch(storage) {
+        case AF_SPARSE_COO: create_sparse_array_from_dense2<T, AF_SPARSE_COO>(out, in, nonZeroIdx); break;
+        case AF_SPARSE_CSR: create_sparse_array_from_dense2<T, AF_SPARSE_CSR>(out, in, nonZeroIdx); break;
+        case AF_SPARSE_CSC: create_sparse_array_from_dense2<T, AF_SPARSE_CSC>(out, in, nonZeroIdx); break;
+        default: AF_ERROR("Invalid Storage Type", AF_ERR_ARG); // Should never come here
+    }
+}
+
+af_err af_create_sparse_array_from_dense(af_sparse_array *out, const af_array in,
+                                         const af_sparse_storage storage)
+{
+    try {
+        // Checks:
+        // storage is within acceptable range
+        // values is of floating point type
+
+        ArrayInfo info = getInfo(in);
+
+        if(!(storage == AF_SPARSE_COO
+          || storage == AF_SPARSE_CSR
+          || storage == AF_SPARSE_CSC)) {
+            AF_ERROR("Storage type is out of range/unsupported", AF_ERR_ARG);
+        }
+
+        TYPE_ASSERT(info.isFloating());
+
+        af_sparse_t sparse;
+
+        af_array nonZeroIdx = 0; // Yes I know how this looks
+        AF_CHECK(af_where(&nonZeroIdx, in));
+
+        ArrayInfo nNZInfo = getInfo(nonZeroIdx);
+        dim_t nNZ = nNZInfo.elements();
+
+        sparse.storage = storage;
+        sparse.nRows = info.dims()[0];
+        sparse.nCols = info.dims()[1];
+        sparse.nNZ = nNZ;
+
+        switch(info.getType()) {
+            case f32: create_sparse_array_from_dense1<float  >(&sparse, in, nonZeroIdx, storage); break;
+            case f64: create_sparse_array_from_dense1<double >(&sparse, in, nonZeroIdx, storage); break;
+            case c32: create_sparse_array_from_dense1<cfloat >(&sparse, in, nonZeroIdx, storage); break;
+            case c64: create_sparse_array_from_dense1<cdouble>(&sparse, in, nonZeroIdx, storage); break;
+            default: TYPE_ERROR(1, info.getType());
+        }
+
+        // Call the conversion in the backend here
+
+        *out = getSparseHandle(sparse);
+
+        if(nonZeroIdx != 0) AF_CHECK(af_release_array(nonZeroIdx));
+    } CATCHALL;
+
+    return AF_SUCCESS;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Get Functions
+////////////////////////////////////////////////////////////////////////////////
+af_err af_get_sparse_values(af_array *out, const af_sparse_array in)
+{
+    try {
+        af_sparse_t sparse = getSparse(in);
+        *out = sparse.values;
```

Comment:
Hmm. Leave it like this for now. We should think about the proper way to do it. Can you add a "TODO" / "FIXME" in there (as well as features) so we can have a look later?